Config replacers scope clash with config import

alex-dukhno opened the pull request:
This PR contains fix for config replacers and config imports for xml based configuration files (as reported in #15452)

@blazember, could you please take a look? (cc @neilstevenson)

devOpsHazelcast commented:
Can one of the admins verify this patch?

blazember commented on hazelcast/src/main/java/com/hazelcast/config/AbstractXmlConfigBuilder.java:
```diff
@@ -92,46 +92,69 @@ public Iterator getPrefixes(String namespaceURI) {
     }
 
-    private void replaceImportElementsWithActualFileContents(Node root) throws Exception {
+    private void replaceVariables(Node root) throws Exception {
```
This and `traverseChildrenAndReplaceVariables(root)` seem to be duplicate methods. Could this be a leftover?

blazember replied:
Now `traverseChildrenAndReplaceVariables` is an unused method. Please don't forget to remove it in #15824.

blazember commented on hazelcast/src/test/java/com/hazelcast/config/XmlConfigImportVariableReplacementTest.java:
```diff
@@ -137,107 +218,94 @@ public void testImportedConfigVariableReplacement() throws Exception {
+                createFileWithContent("hz1", "xml", ""),
+                createFileWithContent("hz2", "xml", ""),
+                createFileWithContent("hz3", "xml", "")
+        );
```
Nice cleanup :+1:

blazember commented on hazelcast/src/test/java/com/hazelcast/config/test/builders/MapXmlStoreConfigBuilder.java:
```diff
@@ -0,0 +1,63 @@
+        return this;
+    }
+
+    public MapXmlStoreConfigBuilder withInitialMode(String initialMode) {
```
You can use the `MapStoreConfig.InitialLoadMode` enum here.

alex-dukhno commented:
@blazember I added tests to `XmlClientConfigImportVariableReplacementTest`, did some cleanup (I'll make some more in #15824) and rebased over latest master.

Thank you for your review.

blazember commented:
run-lab-run

blazember commented on hazelcast/src/test/java/com/hazelcast/IOUtils.java:
```diff
@@ -0,0 +1,45 @@
+import java.io.IOException;
+import java.util.function.Function;
+
+public class IOUtils {
```
This helper feels to be specific to config, I think it should be under `com.hazelcast.config.helpers`. We have `DeclarativeConfigFileHelper` there for similar reasons. It doesn't use static methods, but I think these methods can go there. Could you please check if we can add them there?

blazember commented:
@alex-dukhno Thank you for your quick response. I added one more comment for consideration. Also, there are a few checkstyle errors, you can check them with 
```
mvn clean install -DskipTests -Pcheckstyle,spotbugs
```
in the project root. Could you please take a look?

blazember commented:
run-lab-run

blazember commented:
Thank you for the fix @alex-dukhno, a really nice PR :+1:
